packages/datetime/src/dateInput.tsx

```diff
@@ -447,16 +471,49 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
             // the page. tabbing forward should *not* close the popover, because
             // focus will be moving into the popover itself.
             this.setState({ isOpen: false });
+        } else if (e.which === Keys.ESCAPE) {
+            this.setState({ isOpen: false });
+            this.inputRef.blur();
         }
         this.safeInvokeInputProp("onKeyDown", e);
     };
 
+    // blur DOM event listener (not React event)
+    private handlePopoverBlur = () => this.handleClosePopover();
+
+    private registerPopoverBlurHandler = () => {
+        if (this.contentRef != null) {
+            // Popover contents are well structured, but the selector will need
+            // to be updated if more focusable components are added in the future
+            const tabbableElements = this.contentRef.querySelectorAll("input, [tabindex]:not([tabindex='-1'])");
+            const numOfElements = tabbableElements.length;
+            if (numOfElements > 0) {
+                const lastPopoverElement = tabbableElements[numOfElements - 1] as HTMLElement;
```

Comment:
why do we need to do all this (keep track of the last popover element and juggle the event handlers)? please leave comments

Reply:
at first I thought this was the last popover element on the page, but it's actually more like `lastElementInPopover`

Reply:
- [ ] Add comment on why we keep track of `lastPopoverElement`
- [ ] rename `lastPopoverElement` -> `lastElementInPopover`